File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/saveable/ModelSaver.java
package com.raizlabs.android.dbflow.sql.saveable;

import android.content.ContentValues;
import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.annotation.ConflictAction;
import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.runtime.NotifyDistributor;
import com.raizlabs.android.dbflow.sql.language.Delete;
import com.raizlabs.android.dbflow.structure.BaseModel;
import com.raizlabs.android.dbflow.structure.ModelAdapter;
import com.raizlabs.android.dbflow.structure.database.DatabaseStatement;
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper;

/**
 * Description: Defines how models get saved into the DB. It will bind values to {@link android.content.ContentValues} for
 * an update, execute a {@link DatabaseStatement}, or delete an object via the {@link Delete} wrapper.
 */
public class ModelSaver<TModel> {

    private static final int INSERT_FAILED = -1;

    private ModelAdapter<TModel> modelAdapter;

    public void setModelAdapter(ModelAdapter<TModel> modelAdapter) {
        this.modelAdapter = modelAdapter;
    }

    public synchronized boolean save(@NonNull TModel model) {
        return save(model, getWritableDatabase(), modelAdapter.getInsertStatement(),
            modelAdapter.getUpdateStatement());
    }

    public synchronized boolean save(@NonNull TModel model, DatabaseWrapper wrapper) {
        return save(model, wrapper, modelAdapter.getInsertStatement(wrapper),
            modelAdapter.getUpdateStatement(wrapper));
    }

    @SuppressWarnings("unchecked")
    public synchronized boolean save(@NonNull TModel model, DatabaseWrapper wrapper,
                                     DatabaseStatement insertStatement,
                                     DatabaseStatement updateStatement) {
        boolean exists = modelAdapter.exists(model, wrapper);

        if (exists) {
            exists = update(model, wrapper, updateStatement);
        }

        if (!exists) {
            exists = insert(model, insertStatement, wrapper) > INSERT_FAILED;
        }

        if (exists) {
            NotifyDistributor.get().notifyModelChanged(model, modelAdapter, BaseModel.Action.SAVE);
        }

        // return successful store into db.
        return exists;
    }

    public synchronized boolean update(@NonNull TModel model) {
        return update(model, getWritableDatabase(), modelAdapter.getUpdateStatement());
    }

    public synchronized boolean update(@NonNull TModel model, @NonNull DatabaseWrapper wrapper) {
        DatabaseStatement insertStatement = modelAdapter.getUpdateStatement(wrapper);
        boolean success = false;
        try {
            success = update(model, wrapper, insertStatement);
        } finally {
            // since we generate an insert every time, we can safely close the statement here.
            insertStatement.close();
        }
        return success;
    }

    @SuppressWarnings("unchecked")
    public synchronized boolean update(@NonNull TModel model, @NonNull DatabaseWrapper wrapper,
                                       @NonNull DatabaseStatement databaseStatement) {
        modelAdapter.saveForeignKeys(model, wrapper);
        modelAdapter.bindToUpdateStatement(databaseStatement, model);
        boolean successful = databaseStatement.executeUpdateDelete() != 0;
        if (successful) {
            NotifyDistributor.get().notifyModelChanged(model, modelAdapter, BaseModel.Action.UPDATE);
        }
        return successful;
    }

    @SuppressWarnings("unchecked")
    public synchronized long insert(@NonNull TModel model) {
        return insert(model, modelAdapter.getInsertStatement(), getWritableDatabase());
    }

    @SuppressWarnings("unchecked")
    public synchronized long insert(@NonNull TModel model, @NonNull DatabaseWrapper wrapper) {
        DatabaseStatement insertStatement = modelAdapter.getInsertStatement(wrapper);
        long result = 0;
        try {
            result = insert(model, insertStatement, wrapper);
        } finally {
            // since we generate an insert every time, we can safely close the statement here.
            insertStatement.close();
        }
        return result;
    }

    @SuppressWarnings("unchecked")
    public synchronized long insert(@NonNull TModel model, @NonNull DatabaseStatement insertStatement,
                                    DatabaseWrapper wrapper) {
        modelAdapter.saveForeignKeys(model, wrapper);
        modelAdapter.bindToInsertStatement(insertStatement, model);
        long id = insertStatement.executeInsert();
        if (id > INSERT_FAILED) {
            modelAdapter.updateAutoIncrement(model, id);
            NotifyDistributor.get().notifyModelChanged(model, modelAdapter, BaseModel.Action.INSERT);
        }
        return id;
    }

    public synchronized boolean delete(@NonNull TModel model) {
        return delete(model, getWritableDatabase());
    }

    @SuppressWarnings("unchecked")
    public synchronized boolean delete(@NonNull TModel model, @NonNull DatabaseWrapper wrapper) {
        DatabaseStatement deleteStatement = modelAdapter.getDeleteStatement(model, wrapper);
        boolean success = false;
        try {
            success = delete(model, deleteStatement, wrapper);
        } finally {
            // since we generate an insert every time, we can safely close the statement here.
            deleteStatement.close();
        }
        return success;
    }

    @SuppressWarnings("unchecked")
    public synchronized boolean delete(@NonNull TModel model, @NonNull DatabaseStatement databaseStatement,
                                       @NonNull DatabaseWrapper wrapper) {
        modelAdapter.deleteForeignKeys(model, wrapper);

        boolean success = databaseStatement.executeUpdateDelete() != 0;
        if (success) {
            NotifyDistributor.get().notifyModelChanged(model, modelAdapter, BaseModel.Action.DELETE);
        }
        modelAdapter.updateAutoIncrement(model, 0);
        return success;
    }

    protected DatabaseWrapper getWritableDatabase() {
        return FlowManager.getDatabaseForTable(modelAdapter.getModelClass()).getWritableDatabase();
    }

    public ModelAdapter<TModel> getModelAdapter() {
        return modelAdapter;
    }

    /**
     * @see #save(Object, DatabaseWrapper, DatabaseStatement, DatabaseStatement)
     */
    @Deprecated
    @SuppressWarnings("unchecked")
    public synchronized boolean save(@NonNull TModel model, DatabaseWrapper wrapper,
                                     DatabaseStatement insertStatement, ContentValues contentValues) {
        boolean exists = modelAdapter.exists(model, wrapper);

        if (exists) {
            exists = update(model, wrapper, contentValues);
        }

        if (!exists) {
            exists = insert(model, insertStatement, wrapper) > INSERT_FAILED;
        }

        if (exists) {
            NotifyDistributor.get().notifyModelChanged(model, modelAdapter, BaseModel.Action.SAVE);
        }

        // return successful store into db.
        return exists;
    }

    /**
     * @see #update(Object, DatabaseWrapper, DatabaseStatement)
     */
    @Deprecated
    @SuppressWarnings("unchecked")
    public synchronized boolean update(@NonNull TModel model, @NonNull DatabaseWrapper wrapper,
                                       @NonNull ContentValues contentValues) {
        modelAdapter.saveForeignKeys(model, wrapper);
        modelAdapter.bindToContentValues(contentValues, model);
        boolean successful = wrapper.updateWithOnConflict(modelAdapter.getTableName(), contentValues,
            modelAdapter.getPrimaryConditionClause(model).getQuery(), null,
            ConflictAction.getSQLiteDatabaseAlgorithmInt(modelAdapter.getUpdateOnConflictAction())) != 0;
        if (successful) {
            NotifyDistributor.get().notifyModelChanged(model, modelAdapter, BaseModel.Action.UPDATE);
        }
        return successful;
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/saveable/ModelSaver.kt
package com.raizlabs.android.dbflow.sql.saveable

import android.content.ContentValues
import com.raizlabs.android.dbflow.annotation.ConflictAction
import com.raizlabs.android.dbflow.runtime.NotifyDistributor
import com.raizlabs.android.dbflow.structure.BaseModel
import com.raizlabs.android.dbflow.structure.ModelAdapter
import com.raizlabs.android.dbflow.structure.database.DatabaseStatement
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper

/**
 * Description: Defines how models get saved into the DB. It will bind values to [DatabaseStatement]
 * for all CRUD operations as they are wildly faster and more efficient than [ContentValues].
 */
open class ModelSaver<T : Any> {

    lateinit var modelAdapter: ModelAdapter<T>

    @Synchronized
    fun save(model: T,
             wrapper: DatabaseWrapper): Boolean {
        return save(model, wrapper, modelAdapter.getInsertStatement(wrapper),
                modelAdapter.getUpdateStatement(wrapper))
    }

    @Synchronized
    fun save(model: T,
             wrapper: DatabaseWrapper,
             insertStatement: DatabaseStatement,
             updateStatement: DatabaseStatement): Boolean {
        var exists = modelAdapter.exists(model, wrapper)

        if (exists) {
            exists = update(model, wrapper, updateStatement)
        }

        if (!exists) {
            exists = insert(model, insertStatement, wrapper) > INSERT_FAILED
        }

        if (exists) {
            NotifyDistributor().notifyModelChanged(model, modelAdapter, BaseModel.Action.SAVE)
        }

        // return successful store into db.
        return exists
    }

    @Synchronized
    fun update(model: T, wrapper: DatabaseWrapper): Boolean {
        val updateStatement = modelAdapter.getUpdateStatement(wrapper)
        val success: Boolean
        try {
            success = update(model, wrapper, updateStatement)
        } finally {
            // since we generate an insert every time, we can safely close the statement here.
            updateStatement.close()
        }
        return success
    }

    @Synchronized
    fun update(model: T, wrapper: DatabaseWrapper,
               databaseStatement: DatabaseStatement): Boolean {
        modelAdapter.saveForeignKeys(model, wrapper)
        modelAdapter.bindToUpdateStatement(databaseStatement, model)
        val successful = databaseStatement.executeUpdateDelete() != 0L
        if (successful) {
            NotifyDistributor().notifyModelChanged(model, modelAdapter, BaseModel.Action.UPDATE)
        }
        return successful
    }

    @Synchronized open fun insert(model: T, wrapper: DatabaseWrapper): Long {
        val insertStatement = modelAdapter.getInsertStatement(wrapper)
        var result: Long = 0
        try {
            result = insert(model, insertStatement, wrapper)
        } finally {
            // since we generate an insert every time, we can safely close the statement here.
            insertStatement.close()
        }
        return result
    }

    @Synchronized open fun insert(model: T,
                                  insertStatement: DatabaseStatement,
                                  wrapper: DatabaseWrapper): Long {
        modelAdapter.saveForeignKeys(model, wrapper)
        modelAdapter.bindToInsertStatement(insertStatement, model)
        val id = insertStatement.executeInsert()
        if (id > INSERT_FAILED) {
            modelAdapter.updateAutoIncrement(model, id)
            NotifyDistributor().notifyModelChanged(model, modelAdapter, BaseModel.Action.INSERT)
        }
        return id
    }

    @Synchronized
    fun delete(model: T, wrapper: DatabaseWrapper): Boolean {
        val deleteStatement = modelAdapter.getDeleteStatement(wrapper)
        var success = false
        try {
            success = delete(model, deleteStatement, wrapper)
        } finally {
            // since we generate an insert every time, we can safely close the statement here.
            deleteStatement.close()
        }
        return success
    }

    @Synchronized
    fun delete(model: T,
               deleteStatement: DatabaseStatement,
               wrapper: DatabaseWrapper): Boolean {
        modelAdapter.deleteForeignKeys(model, wrapper)
        modelAdapter.bindToDeleteStatement(deleteStatement, model)

        val success = deleteStatement.executeUpdateDelete() != 0L
        if (success) {
            NotifyDistributor().notifyModelChanged(model, modelAdapter, BaseModel.Action.DELETE)
        }
        modelAdapter.updateAutoIncrement(model, 0)
        return success
    }

    /**
     * Legacy save method. Uses [ContentValues] vs. the faster [DatabaseStatement] for updates.
     *
     * @see .save
     */
    @Deprecated("")
    @Synchronized
    fun save(model: T,
             wrapper: DatabaseWrapper,
             insertStatement: DatabaseStatement,
             contentValues: ContentValues): Boolean {
        var exists = modelAdapter.exists(model, wrapper)

        if (exists) {
            exists = update(model, wrapper, contentValues)
        }

        if (!exists) {
            exists = insert(model, insertStatement, wrapper) > INSERT_FAILED
        }

        if (exists) {
            NotifyDistributor().notifyModelChanged(model, modelAdapter, BaseModel.Action.SAVE)
        }

        // return successful store into db.
        return exists
    }

    /**
     * @see .update
     */
    @Deprecated("")
    @Synchronized
    fun update(model: T, wrapper: DatabaseWrapper, contentValues: ContentValues): Boolean {
        modelAdapter.saveForeignKeys(model, wrapper)
        modelAdapter.bindToContentValues(contentValues, model)
        val successful = wrapper.updateWithOnConflict(modelAdapter.tableName, contentValues,
                modelAdapter.getPrimaryConditionClause(model).query, null,
                ConflictAction.getSQLiteDatabaseAlgorithmInt(modelAdapter.updateOnConflictAction)) != 0L
        if (successful) {
            NotifyDistributor().notifyModelChanged(model, modelAdapter, BaseModel.Action.UPDATE)
        }
        return successful
    }

    companion object {

        val INSERT_FAILED = -1
    }
}

